BinaryTypeBinding: enrich RuntimeException with filename

jukzi opened the pull request:
During internal errors it is not clear which file did contain the problematic binary. For example:

https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3516

Example output will be like: "java.lang.RuntimeException: RuntimeException caching =P/libGh375.jar|TestGh375.class"

stephan-herrmann commented:
If this is just for investigating #3516 and will be removed afterwards, I have no objections.

If, however, this is intended to stay, wouldn't it be better to wrap the exception in or near BTB.cachePartsFrom()? That would benefit more scenarios, no?

As for the exception message: it's not really the caching that creates problems, but loading the binary type does :)

jukzi commented:
> If, however, this is intended to stay, wouldn't it be better to wrap the exception in or near BTB.cachePartsFrom()? That would benefit more scenarios, no?

I would like to let it stay as similar error may appear in the future. I moved it to BTB as requested and changed the text to "loading". ok?
